### ISSUE TITLE
Create a fully "modifiable" nvim terminal emulator

### ISSUE THREAD

**Issue by egr95** [neovim/neovim#23645]
### Problem

The Neovim terminal ([nvim_terminal_emulator](https://neovim.io/doc/user/nvim_terminal_emulator.html)) is awesome. Combined with plugins such as [vim-slime](https://github.com/jpalardy/vim-slime), it enables powerful IDE-like functionality.

I think the Nvim terminal would be even better if it allowed for a fully "modifiable" mode, more akin to the [Emacs terminal](https://www.gnu.org/software/emacs/manual/html_node/emacs/Terminal-emulator.html). What I mean by this, is a shell running inside of neovim with a *separate* terminal buffer, such that this buffer can be manipulated like a regular vim buffer (including normal mode) as well as allowing the user to send commands to the underlying shell (and updating the buffer with any output).

As far as I know, the current Nvim terminal does not allow for any normal mode edits to the command prompt. For example, pressing `dw` in a terminal buffer will not work (if one has `:set modifiable`, then `dw` will work on any text *prior to the command prompt*, but not on the command prompt itself, where it is most needed). Likewise, the normal mode and terminal mode cursors are distinct, meaning that if I exit from terminal mode to normal mode, move the cursor backwards by a couple of words, and enter terminal mode again, the cursor will return to the position from which I exited terminal mode, essentially "forgetting" any movement that happened in normal mode. 

I do realize that I can activate vi mode in the terminal itself, and this is indeed my current workflow. However, this is suboptimal since it requires navigating through 2 different normal modes, the Nvim normal mode and the terminal vi normal mode, which are not exactly the same and can throw off one's muscle memory.

### Expected behavior

It would be awesome if `:set modifiable` made the terminal behave more like any other Nvim buffer, with fully integrated normal and terminal modes. In other words, it should allow for arbitrary normal mode commands, including on the command prompt. So if I exit to normal mode, press `dw`, and return to terminal mode, a word should have been deleted. In addition, if I exit to normal mode, press `b`, and return to terminal mode, the cursor should now stand one word back from the original position (note this would allow effective navigation of the command prompt using plugins such as [easymotion](https://github.com/easymotion/vim-easymotion)). When in terminal mode, pressing Enter should still allow me to execute commands, and the buffer should automatically update with any new lines from the terminal output.

**Comment by egr95** [comment]
Linking to a similar proposal from a few years back: #8805

**Comment by justinmk** [comment]
I'm still thinking the approach in https://github.com/neovim/neovim/issues/8805#issuecomment-409940181 might be best:

> start with a basic adapter like:
> 
> * copy the last line of the buffer
> * delete the shell prompt line: send `ctrl-u` or `<bs><bs><bs>...` (or `ctrl-c` in cmd.exe)
>   
>   * guess the shell prompt "prefix" by comparing the state after clearing the prompt line
> * send the user's edited line back to the terminal job (shell)

**Comment by nyngwang** [comment]
> [...] such that the buffer becomes more like a regular vim buffer (including normal mode commands)

All lines above the current prompt line are history, which is the output of commands that have been executed. So treating it as a regular Vim buffer would be a little strange since there is no point to change the log/output of sent commands. So I more like the way quoted by justinmk (but I'm still thinking about whether making it a floating window would be cumbersome(?) for users)

Regarding editing the current (possibly multi-line) command, I fully support the idea.

Just some opinions.

**Comment by egr95** [comment]
Agreed it would be much more important to enable normal-mode edits on the command prompt than on the history (the current Nvim terminal unfortunately only allowing for the latter but not the former).

I am not familiar with the Nvim internals, but the approach proposed by justinmk seems very reasonable to me. Not sure if any additional tweaking would be needed in order to support multi-line commands as well.

**Comment by egr95** [comment]
Adding a link to a plugin that I just came across, which provides a neat solution to some of the above issues: [term-edit.nvim](https://github.com/chomosuke/term-edit.nvim).

**Comment by brookhong** [comment]
Maybe we just need two APIs for this,

* nvim_term_get_input(terminal_job_id)
* nvim_term_set_input(terminal_job_id, input)

**Comment by justinmk** [comment]
@brookhong maybe, but that would be done as a channel API, not terminal-specific I think. And it's already possible to listen to the terminal channel given by the `&channel` buffer-local option.

Actually, how is your proposal going to solve this? The handler would need to understand the underlying process. Did you try `buftype=prompt` buffers?

**Comment by brookhong** [comment]
Just had a try by running zsh in a prompt buffer. Command line editing almost works with some exceptions such as `u`, but it does not work perfectly as a real terminal, since it needs a thorough callback function to handle I/O. Prompt buffer can be a good choice for prompt implementation of customized/lightweight shells in Neovim, but it does not work well with a complete external shell like bash/zsh at least for now.

For terminal buffer, I also prefer your approach above:

> * copy the last line of the :terminal buffer to a new, temporary, editable floating window
> * delete the shell prompt line: send `ctrl-u` or `<bs><bs><bs>...` (or `ctrl-c` in cmd.exe)
>
>   * guess the shell prompt "prefix" by comparing the state after clearing the prompt line
> * send the user's edited line back to the terminal job (shell)

But it's not enough to copy only the last line, since for a long command input, it will be split into multiple lines in a terminal buffer. That's why we need an API `nvim_term_get_input` to get input, which can also avoid the trouble in Lua of handling different prompts based on user preferences / from different shells.

To set user edited content back to the terminal, currently we can use `vim.fn.chansend(terminal_job_id, "\x15"..user_edited_content)`, but I'm not sure `\x15` (which is `Ctrl-u`) works well for all shells. That's why we need `nvim_term_set_input`.

**Comment by justinmk** [comment]
> since it needs a thorough callback function to handle I/O. ... but it does not work well with a complete external shell like bash/zsh at least for now.

How is that situation going to change in a terminal buffer running a shell process? `nvim_term_get_input` doesn't know anything about how the underlying shell (or whatever) process treats the input.

> But it's not enough to copy only the last line, since for a long command input, it will be split into multiple lines in a terminal buffer.

I did not say the last line.

Anyway, this can probably be handled by emitting some sort of OSC sequence.

**Comment by TheLeoP** [comment]
I've been iterating on top of https://github.com/xb-bx/editable-term.nvim (mentioned in https://github.com/neovim/neovim/issues/8563#issuecomment-2802019894) with the intention of eventually opening a PR to Neovim in order to upstream it, and I've found some rough edges with the approach, even though it's almost all the way there. So, I wanted to receive some feedback on my proposal/findings before opening a PR proper.

My current version of `editable-term.lua` is https://github.com/TheLeoP/nvim-config/blob/bae7e8cd7eb6220e92174d7507145a465d562249/lua/personal/editable-term.lua#L1-L278 and all of my finding are on comments on that same file, some of the most significant are

> Intercepting the `c` operator in `ModeChanged` does not work with empty regions (e.g. `ci"` with the text `grep ""`) because the marks `[` and `]` return {0, 0}. TextYankPost can be used to compute the change ourselves and make this work with empty regions, but it wouldn't work when using the blackhole register.

> The problem with empty ranges includes plugins workarounds like on mini.ai or targets.vim, but for a different reason. Their workaround is to insert a single space in the empty region, visually select it and allow only operators like `c` or `d` (because they'll be able to delete the space)

> `TextChanged` gets retriggered by the changes made on `chansend()`, so we need to ignore it for a small amount of time

> `TextChanged` gets triggered on terminal reflow, which may happen when opening a new window on a split or even on `q:`

My previous version of `editable-term.lua` (https://github.com/TheLeoP/nvim-config/blob/b07e6593afb9ba1d13b89e1cb1090d51f35a9dfc/lua/personal/editable-term.lua#L1-L266) didn't handle correctly `c` nor `d` when using the blackhole register, but it did handle empty regions with the `c` operator because it was computing the change itself inside of the `TetxYankPost` autocmd.

Another finding that I forgot to add to the file is that some shells have some kind of autosuggestion behavior where additional text will be shown in the prompt while typing (usually in a grayish shade), but will be discarded unless the user explicitly accepts it. Since Neovim has no way of distinguishing between the command typed by the user and the text suggested by the shell, my approach has been to add a configuration for a key to clean the autosuggestion and default it to an empty string.

My main modifications in top of the original editable-term.nvim are

- correct handling of multibyte chars
- handling prompts spanning multiple visual lines
- handling changes made while using the blackhole register
- drop support for pattern based prompt matching (at least in a first implementation) to simplify the code, relying solely in the OSC133 escape sequence (which, btw, seems to use `utf` indexing instead of `byte` indexing)

I'm open to any kind of feedback, but I'll also try to create a PR later this week.

@justinmk

**Comment by justinmk** [comment]
That seems like more evidence that [this approach](https://github.com/neovim/neovim/issues/23645#issuecomment-1576850696) is the most sane. Then the only "messy part" is (1) getting the line from the shell, and (2) updating/overwriting it. The ctrl-u approach is the "dumb" solution, and can always be replaced with a smart solution later. No other magic or hacks are needed, just show a normal buffer in a floatwin, then some keymap "submits" it.

**Comment by kalomak** [comment]
> That seems like more evidence that [this approach](https://github.com/neovim/neovim/issues/23645#issuecomment-1576850696) is the most sane. Then the only "messy part" is (1) getting the line from the shell, and (2) updating/overwriting it. The ctrl-u approach is the "dumb" solution, and can always be replaced with a smart solution later. No other magic or hacks are needed, just show a normal buffer in a floatwin, then some keymap "submits" it.

I attempted this read/write API + floating window approach. 
https://github.com/kalomak/termio.nvim (has demo videos)

The editor is pretty complex but the API is simple and can be hooked up to a very minimal window.

It currently has some major issues like:
```zsh
>> ls<tab>
ls            lsappinfo     lsbom         lskq          lsm           lsmp          lsof          lstmeval      lstmtraining  lsvfs
```
These completions are included when `read()` is called, there is no way to differentiate them (without regex at least) from the actual user typed command.  This should be solvable, VSCode handles this by setting up custom markers (if I understood correctly) [here](https://github.com/microsoft/vscode/blob/621079ff563d3d38b7c167b3d74425c629d0584f/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L351-L377). 

> EDIT: the plugin now clears completions before reading via shell integration.

So a lot remains to be done but I think this is already an improvement over the standard terminal experience.

**Comment by justinmk** [comment]
> VSCode handles this by setting up custom markers (if I understood correctly) [here](https://github.com/microsoft/vscode/blob/621079ff563d3d38b7c167b3d74425c629d0584f/src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh#L351-L377).

See `:help shell-prompt`

**Comment by kalomak** [comment]
> See `:help shell-prompt`

Yeah unfortunately there is no standard marker for tab completions. So need to either wrap shell hooks to print custom OSC sequences like vscode does or manage a marker inside neovim.

**Comment by kalomak** [comment]
Update on the read / write API  + floatwin approach:

TLDR: There are ways to make this work in most cases but afaik not in all cases.

It is pretty straightforward to create a clean and fast read / write integration with a command-line editor, like zle for zsh, readline for bash. This can be achieved e.g. via a named pipe. However, this will only work with the particular process it is set up for, so e.g. activating a REPL or sshing will break this connection. This would be the most native feeling, but most narrow solution.

Another slightly less clean way is to use OSC markers, which neovim already supports, to infer where the command starts and reading from the terminal buffer. This is dependent on the user having set up OSC markers correctly. Most [terminal emulators do this for your shell](https://github.com/ghostty-org/ghostty/blob/main/src/shell-integration/zsh/ghostty-integration) by default, but most REPLs do not. This also runs into some problems like parsing what part of the terminal buffer is the command and what is a suggestion / completion, which is surprisingly hard. 

There is a way that is not dependent on shell integration or markers:
- send C-e (goes to command end)
- mark cursor position as command end (needed to ignore completions and suggestions)
- read + cache terminal buffer content
- send C-u (clears the command) or C-a if clearing the command here is not desired
- mark cursor position as command start
- set the text between command start and end to the floatwin

This seems to work in all shells and REPLs I have tried UNLESS a 'vi mode' is enabled in the shell editor. Also [only around 300 LOC](https://github.com/kalomak/termio.nvim/blob/e11767af44581c70ead3342affc0312eee58ff50/lua/termio/init.lua). But, even this fails in some cases, for example if the command is larger than the terminal window.

Write is easy with all approaches. `nvim_chan_send` with bracketed paste is fast and reliable. '\r' at the end submits.

If the API works it is also relatively easy to extend the floatwin approach to making the terminal buffer command area itself editable. It is already possible to modify the terminal buffer independent of the terminal state and it does not seem to cause issues.  Then the command part of the desynced terminal buffer could be treated the same way as the floatwin editor. 

If you are interested, you can try prototypes for these with [termio](https://github.com/kalomak/termio.nvim) (or just check demo videos):
- `overlay` floatwin on top of the command
- `centered` centered floatwin
- `integrated` editable terminal buffer, hacky

These rely on manually sourcing a shell integration script and OSC markers. README tells how,`:checkhealth termio` checks if you have them set up.
